src/lib/pure_pursuit/PurePursuit.hpp

```diff
@@ -0,0 +1,83 @@
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include <matrix/math.hpp>
+
+using namespace matrix;
+
+/**
+ * @file PurePursuit.hpp
+ *
+ * Implementation of pure pursuit guidance logic.
+ *
+ * Pure pursuit is a path following algorithm that uses the intersection between the path and
+ * a circle (the radius of which is referred to as lookahead distance) around the vehicle as
+ * the target point for the vehicle.
+ * 							C
+ * 		  				       /
+ * 						    __/__
+ * 						  /  /    \
+ * 						 /  /      \
+ * 						|  /  V     |
+ * 						 \/        /
+ * 						 /\ _____ /
+ * 	         	    N (0 rad)		/
+ * 			        ^	       P
+ * 				|
+ * 				| D
+ *  	   (-1.5708 rad) <----- ⨂ -----> E (1.5708 rad)
+ * 				|
+ * 				|
+ * 				⌄
+ * 			(+- 3.14159 rad)
+ *
+ * Using the current/prev waypoint and the vehicle position in NED frame as well as the lookahead distance,
+ * this intersection point is calculated and the heading towards it is returned.
+ */
+class PurePursuit
+{
+public:
+	/**
+	 * @brief Return heading towards the intersection point between a circle with a radius of
+	 * lookahead_distance around the vehicle and an extended line segment from the previous to the current waypoint.
+	 * Will fallback to returning the bearing to the current waypoint if inputs are invalid or if no
+	 * intersection points exist.
+	 * Exception: Will return 0 if curr_wp_ned or curr_pos_ned are invalid.
+	 * @param curr_wp_ned Current waypoint in NED frame.
+	 * @param prev_wp_ned Previous waypoint in NED frame.
+	 * @param curr_pos_ned Current position of the vehicle in NED frame.
```

Comment:
Units in meters?

Reply:
Yes, added clarification in [54e99ae](https://github.com/PX4/PX4-Autopilot/pull/23387/commits/54e99ae35fcc2075d8e393ccf89084c19b1b509d).